Fix text kerning calculations and some FT2Font cleanup

QuLogic opened the pull request:
## PR Summary

Kerning values were incorrectly divided by 64 (possibly some confusion from FreeType using both 26.6 and 16.16 types) rendering them effectively 0, unless used with very large fonts.

This corrects that error, but also adds an rcParam (`text.kerning_factor`) to restore the old behaviour. This is set to the old value in the `_classic_test` style to allow ~100 test images to continue working. There are about 20 more tests using the new style that fail now. If this looks like a reasonable way to write out the rcParam, I will add a line in those tests so their images do not need to be regenerated.

Fixes #14887.

## PR Checklist
- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

anntzer commented on lib/matplotlib/font_manager.py:
```diff
@@ -1285,7 +1285,8 @@ def is_opentype_cff_font(filename):
         hinting_factor = rcParams['text.hinting_factor']
-    return _get_font(os.fspath(filename), hinting_factor)
+    return _get_font(os.fspath(filename), hinting_factor,
+                     _kerning_factor=rcParams['text.kerning_factor'])
```
What about folding this into `rcParams['_internal.classic_mode']`?  (how many images would need to be regen'd?)

QuLogic replied:
Well, I didn't update the non-classic tests yet, so you can see from the CI results it'd be 21 failures.

anntzer replied:
Had a quick look at the failures.  A few of them can be improved:
- regen test_acorr, test_image_interps without any text (remove_text)
- test_titletwiny can be rewritten to a non-image-comparison, just checking the title position, like test_titlesetpos which is just below it
- for test_hatching the text labels can probably removed? (can we generate legends with empty labels? not sure...)
- I think the others images would need to be regen'd (assuming we fold the switch into _internal.classic_mode) and need to keep the text to be meaningful.

Or perhaps make this a private rcParam (leading underscore, like _internal.classic_mode)?  Or perhaps I'm just overthinking this and text.kerning_factor is fine.

QuLogic replied:
Well, I added the rcParam for downstream testing's benefits; not sure if only classic mode will be helpful there.

anntzer replied:
Sure.  I guess my preferred solution would be _internal.old_kerning or something like that, but won't insist on it.

QuLogic replied:
I opened #15054 to track these suggestions, but for tests here I just modified the `rcParam` to use the old value.

jklymak commented:
Does this need further discussion?  How many images will have to be rebuilt, and can this be done at the same time as other tests that need an image rebuild?

QuLogic commented:
The only question remaining is what to name the rcParam. With that, we can allow classic tests to continue working (~100 tests) and then the remaining ones can patch the parameter to avoid regenerating results (16 tests).

anntzer commented:
Given that there's already text.hinting_factor which should be considered a relic of the past, text.kerning_factor looks fine to me.

QuLogic commented:
Rebased for fixed docs and added a [what's new entry](https://24310-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/users/next_whats_new/2019-08-14-ES.html).

QuLogic commented:
@tacaswell any reason to wait on this now that it's passing?

anntzer commented:
Seems like a pity to wait for 3.3 so I remilestoned this to 3.2.  @tacaswell please ping if you think this should wait, otherwise I'll merge.

anntzer commented on src/ft2font_wrapper.cpp:
```diff
@@ -553,10 +553,11 @@ static int PyFT2Font_init(PyFT2Font *self, PyObject *args, PyObject *kwds)
 
     if (!PyArg_ParseTupleAndKeywords(
-             args, kwds, "O|l:FT2Font", (char **)names, &fname, &hinting_factor)) {
+             args, kwds, "O|li:FT2Font", (char **)names, &fname, &hinting_factor, &kerning_factor)) {
```
do you want to make _kerning_factor keyword-only? (you can use the "$" specifier for that) otherwise it effectively becomes public API as you can pass it positionally...

QuLogic replied:
Sure, let's do that.
